zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProvider.java

```diff
@@ -172,14 +143,95 @@ public boolean isValid(String id) {
     }
   }
 
+  /**
+   * Initialize a new ClientUriDomainMappingHelper instance if no one has been instantiated for the ACL provider.
+   * @param zks
+   */
   private ClientUriDomainMappingHelper getUriDomainMappingHelper(ZooKeeperServer zks) {
     if (uriDomainMappingHelper == null) {
       synchronized (this) {
         if (uriDomainMappingHelper == null) {
-          uriDomainMappingHelper = new ZkClientUriDomainMappingHelper(zks);
+          ZkClientUriDomainMappingHelper helper = new ZkClientUriDomainMappingHelper(zks);
+          // Set up AuthInfo updater to refresh connection AuthInfo on any client domain changes.
+          helper.setDomainAuthUpdater((cnxn, clientUriToDomainNames) -> {
```

Comment:
I think the intention we want to show here is that the URI-domain mapping doesn't necessarily need to be stored inside ZK. This anonymous class implementation is clearly a ZK-based implementation.

If we had something like

`helper.setDomainAuthUpdater(zkBasedDomainAuthUpdater);`, then it would be easier and more straightforward to change it to
`helper.setDomainAuthUpdater(externalSourceBasedDomainAuthUpdater);`

or potentially make this configurable with Reflection in the future.

Reply:
I see your point. My argument is that this specific design is for ZK only. It would be better if I can put all the related logic to ZkClientUriDomainMappingHelper instead of X509ZNodeGroupAclProvider. However, that will require me to pass a good amount of variables to the helper instance and there will be race condition concerns. The current design assumes we have only one implementation. At the same time, trying to keep the ClientUriDomainMappingHelper interface intact. Once we have an additional external source, then we just need to refactor the implementations instead of the interface.

On the other hand, I'm not convinced that we need an updater for every possible external data source. So I don't like the idea of adding this concept to ClientUriDomainMappingHelper.
For instance, the other external source may not provide a watcher/callback mechanism, so polling is the way to go. In this case, getDomains and updateDomainBasedAuthInfo (for repeatedly updating the AuthInfo) are good enough. And there is nothing we can define about the updater.

Alternatively, we can refactor ClientUriDomainMappingHelper to a pure updater interface. So no more getDomains but only update or register customized update requests. It could be a cleaner design.
If we decide to go this route though, I would suggest separating this change to another refactoring PR. It is not a must for now.

Reply:
I tried to separate this anonymous method out. The good news is that there are no additional concurrency control requirements. The bad news is that we will need to pass some X509ZNodeGroupAclProvider specific info to the new class. The ugliest ones are the target schemes (super and X509). If it is just passing a list of shemes to the updater, then it is still generic enough. However, super is specially handled so it will eventually require us to pass a "super user scheme" + "all the other schemes". I think it is overkill for now. And without the other use cases, it is not possible to do the separation right.

I will not proceed to separate this anonymous class. Adding a todo here.